File: src/note.ts
// A note is one .md file: `---` YAML frontmatter + markdown body. Files are
// truth, so parsing never throws — a note with broken frontmatter still indexes
// (whole file as body) and carries a warning flag for `doctor` to report.
import { basename } from "node:path";

export type Note = {
  /** vault-relative path — the note's identity */
  path: string;
  /** filename stem — what a bare `[[stem]]` link resolves against */
  slug: string;
  title: string;
  type: string | undefined;
  frontmatter: Record<string, unknown>;
  body: string;
  /**
   * Deduped wikilink targets as written, in first-seen order: a bare filename
   * stem (`[[acme]]`) or a vault-relative path without its extension
   * (`[[customers/acme]]`). Resolution is `resolveEdges`' job — a target is
   * matched against the indexed note paths, never joined onto the filesystem.
   */
  links: string[];
  /**
   * sha256 (hex) of `raw` re-encoded as UTF-8 — the decoded string, not the
   * bytes on disk, so an encoding quirk the decoder normalizes away is
   * invisible here. It is a dirtiness check for reindexing, not an identity:
   * identity is `path`.
   */
  hash: string;
  /**
   * Frontmatter the parser could not use as written: unterminated or malformed
   * block, non-mapping value, alias expansion over budget, or a non-string
   * `title`/`type`. The note still indexes; `doctor` reports the flag.
   */
  malformedFrontmatter: boolean;
};

/**
 * The path-qualified wikilink target of a vault-relative path: the path minus
 * `.md`. What the write gate links with and what `doctor` offers as the fix for
 * an ambiguous link, so both say the same thing. A note at the vault root has
 * no qualified form — its path minus `.md` *is* its bare stem.
 */
export const linkTarget = (rel: string): string => rel.slice(0, -".md".length);

/**
 * Can this link target survive a round trip through `[[…]]`? A `[`, `]` or `|`
 * is legal in a filename but destroys or misdirects the link when spliced into
 * one (`[[[archive]/acme]]` parses as no edge at all; `[[my|dir/acme]]` parses
 * as link `my` with an alias), and `linkOf` trims, so leading or trailing
 * whitespace does not survive either (`[[ archive/acme]]` reads back as a
 * different note's path). A rewrite that cannot round-trip is never made.
 */
export const isWritableTarget = (target: string): boolean =>
  !/[\[\]\r\n]/.test(target) && linkOf(target) === target;

// The one wikilink scanner: `parseNote` finds edges with it and `qualifyLinks`
// rewrites with it, so what counts as a link and what gets rewritten cannot
// disagree. Single-line and length-capped: a stray `[[` must not pair with a
// `]]` pages later and invent an edge.
const WIKILINK = /\[\[([^\[\]\n]{1,256})\]\]/g;

/** A match's link target: the text before any `|alias`, trimmed. */
const linkOf = (inner: string): string => inner.split("|")[0]!.trim();

/** Split a leading `---` block off the top of the file. */
function splitFrontmatter(text: string): {
  yaml: string | null;
  body: string;
  unterminated: boolean;
} {
  const none = { yaml: null, body: text };
  if (!text.startsWith("---\n")) return { ...none, unterminated: false };
  const rest = text.slice(4);
  const close = rest.search(/^---[ \t]*$/m);
  // No closing fence (or one with trailing junk): keep the whole file as body,
  // but flag it — a mangled fence would silently drop superseded_by otherwise.
  if (close === -1) return { ...none, unterminated: true };
  return {
    yaml: rest.slice(0, close),
    body: rest.slice(close).replace(/^---[ \t]*\n?/, ""),
    unterminated: false,
  };
}

/**
 * YAML aliases re-expand on every reference, so a 250-byte billion-laughs note
 * parses into megabytes. Count every visit (shared refs accumulate) and bail
 * once the object is too big to belong in an index row.
 */
function withinNodeBudget(value: unknown, budget = 4096): boolean {
  let seen = 0;
  const visit = (v: unknown): boolean => {
    if (++seen > budget) return false;
    if (v && typeof v === "object") {
      for (const child of Object.values(v)) if (!visit(child)) return false;
    }
    return true;
  };
  return visit(value);
}

/**
 * A frontmatter block's YAML as a mapping we can index, or null when it is
 * unusable: invalid YAML, not a mapping, or over the alias budget. Both readers
 * of a block — `parseNote` and the textual patcher below — decide through this
 * one function, so they can never disagree about whether a block is real. (They
 * did once: a fenced-but-invalid block looked patchable while `parseNote`
 * ignored it, so a `superseded_by` written into it was never read.)
 */
function usableFrontmatter(yaml: string): Record<string, unknown> | null {
  if (yaml.trim() === "") return {};
  try {
    const parsed = Bun.YAML.parse(yaml);
    // Must be a mapping, and small enough to store in an index row.
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) && withinNodeBudget(parsed)
      ? (parsed as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}

export function parseNote(raw: string, relPath: string): Note {
  const hash = new Bun.CryptoHasher("sha256").update(raw).digest("hex");
  const lf = raw.replaceAll("\r\n", "\n");
  const text = lf.charCodeAt(0) === 0xfeff ? lf.slice(1) : lf; // strip UTF-8 BOM
  const slug = basename(relPath, ".md");

  const { yaml, body: split, unterminated } = splitFrontmatter(text);
  let frontmatter: Record<string, unknown> = {};
  let body = text;
  let malformedFrontmatter = unterminated;
  if (yaml !== null) {
    const parsed = usableFrontmatter(yaml);
    if (parsed === null) malformedFrontmatter = true;
    else {
      frontmatter = parsed;
      body = split;
    }
  }

  const fmTitle = frontmatter["title"];
  const fmType = frontmatter["type"];
  // Present but not a string: unusable, and doctor should hear about it.
  if ((fmTitle !== undefined && typeof fmTitle !== "string") ||
      (fmType !== undefined && typeof fmType !== "string")) {
    malformedFrontmatter = true;
  }

  // ponytail: regex scan, no markdown parse — wikilinks and headings inside code
  // fences are counted (documented MVP simplification); parse structure if it
  // bites. `/` is ordinary content here — a path-qualified target travels whole
  // and is resolved as SQL, never as a filesystem path.
  const links = [
    ...new Set(Array.from(body.matchAll(WIKILINK), (m) => linkOf(m[1]!))),
  ].filter((s) => s !== "");

  return {
    path: relPath,
    slug,
    title:
      (typeof fmTitle === "string" && fmTitle.trim() !== "" ? fmTitle : undefined) ??
      body.match(/^#[ \t]+(.+?)[ \t]*$/m)?.[1] ??
      slug,
    type: typeof fmType === "string" ? fmType : undefined,
    frontmatter,
    body,
    links,
    hash,
    malformedFrontmatter,
  };
}

/**
 * Render a note we authored. **Never use this to rewrite a human-authored
 * file:** a YAML round-trip is lossy against hand-written data (comments are
 * dropped, `01234` becomes `1234`, `1.0` becomes `1`). Per DESIGN.md § Write
 * gate, edits to existing notes are textual patches of the frontmatter block.
 */
export function serializeNote(note: Pick<Note, "frontmatter" | "body">): string {
  if (Object.keys(note.frontmatter).length === 0) return note.body;
  const yaml = Bun.YAML.stringify(note.frontmatter, null, 2).replace(/\n*$/, "\n");
  return `---\n${yaml}---\n${note.body}`;
}

/**
 * The frontmatter block as *text*: where its YAML starts and ends inside `raw`,
 * and where the body begins after the closing fence. Offsets are into `raw`, so
 * a leading BOM and CRLF endings are stepped over rather than normalized away.
 * Null when `parseNote` would not use the block — no opening fence, no closing
 * fence, or YAML it cannot read — so text-level edits and the parsed view never
 * disagree about what counts as frontmatter.
 */
function frontmatterBlock(
  raw: string,
): { start: number; end: number; bodyStart: number } | null {
  const bom = raw.charCodeAt(0) === 0xfeff ? 1 : 0;
  const open = /^---\r?\n/.exec(raw.slice(bom));
  if (open === null) return null;
  const start = bom + open[0].length;
  const close = /^---[ \t]*\r?(\n|$)/m.exec(raw.slice(start));
  if (close === null) return null;
  const at = {
    start,
    end: start + close.index,
    bodyStart: start + close.index + close[0].length,
  };
  const yaml = raw.slice(at.start, at.end).replaceAll("\r\n", "\n");
  return usableFrontmatter(yaml) === null ? null : at;
}

/**
 * Set one frontmatter key by editing the text, never by re-serializing: append
 * the key's line, or replace the single line that already defines it, leaving
 * every other byte of the file identical. Per DESIGN.md § Write gate this is
 * how the gate touches notes it did not author — a YAML round-trip drops
 * comments and rewrites `01234` and `1.0`. A file with no usable frontmatter
 * block (absent, or malformed enough that `parseNote` ignores it) gets a fresh
 * block prepended; its content is left untouched below.
 *
 * A `null` value **unsets** the key instead — every line defining it goes. A
 * gate-owned key has to be removable, or it would outlive the fact it records
 * (a stale `vault_source` beside a fresh `vault_agent` asserts a pairing that
 * never happened).
 */
export function patchFrontmatter(raw: string, key: string, value: string | null): string {
  // Keys are ours (`superseded_by`, `updated`); anything else could be YAML or
  // regex syntax rather than a key, and this function does not sanitize.
  if (!/^[A-Za-z0-9_-]+$/.test(key)) throw new Error(`patchFrontmatter: unusable key ${key}`);
  const at = frontmatterBlock(raw);
  if (value === null) {
    // Unsetting is only ever a deletion: a file with no usable block does not
    // gain one to hold the absence of a key, and an absent key is already gone.
    if (at === null) return raw;
    const gone = raw
      .slice(at.start, at.end)
      .replace(new RegExp(`^${key}[ \t]*:[^\r\n]*\r?\n?`, "gm"), "");
    return raw.slice(0, at.start) + gone + raw.slice(at.end);
  }
  // A timestamp is a YAML-native scalar and reads as one; anything else is
  // quoted, so a path or a colon can never become syntax.
  const line = `${key}: ${ISO_TIMESTAMP.test(value) ? value : JSON.stringify(value)}`;
  if (at === null) return `---\n${line}\n---\n${raw}`;

  const eol = raw.slice(0, at.start).endsWith("\r\n") ? "\r\n" : "\n";
  const block = raw.slice(at.start, at.end);
  // Top-level only: an indented `key:` belongs to a nested mapping. `[^\r\n]*`
  // rather than `.*` so a CRLF file keeps its carriage return. Every occurrence
  // is taken: YAML's last-wins would otherwise let a duplicate key downstream
  // resurrect the value we just replaced.
  const existing = new RegExp(`^${key}[ \t]*:[^\r\n]*`, "gm");
  let seen = 0;
  const swapped = block.replace(existing, () => (++seen === 1 ? line : ""));
  const patched =
    seen > 0 ? swapped : block + (block === "" || block.endsWith("\n") ? "" : eol) + line + eol;
  return raw.slice(0, at.start) + patched + raw.slice(at.end);
}

/** `2026-08-01T09:41:00.000Z` — what `new Date().toISOString()` produces. */
const ISO_TIMESTAMP = /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\.\d+)?Z$/;

/**
 * Rewrite every bare `[[stem]]` / `[[stem|alias]]` link in `raw`'s body to the
 * path-qualified `target`, leaving every other byte — the frontmatter block,
 * CRLF endings, a BOM — untouched: this edits files the vault did not author.
 * The scan is the parser's own link regex over the same body region `parseNote`
 * reads (`frontmatterBlock` decides both), so what this rewrites and what
 * counts as an edge cannot disagree — which also means links inside code fences
 * are rewritten, the parser's documented MVP simplification. Alias text travels
 * verbatim. Used by the indexer when a stem collision is created (DESIGN.md
 * § Data model).
 */
export function qualifyLinks(raw: string, stem: string, target: string): string {
  const start = frontmatterBlock(raw)?.bodyStart ?? 0;
  // The link regex never spans a line, so a CRLF file's `\r`s all sit outside
  // any match and survive; the frontmatter's bytes are not scanned at all.
  const body = raw.slice(start).replace(WIKILINK, (whole, inner: string) => {
    if (linkOf(inner) !== stem) return whole;
    const pipe = inner.indexOf("|");
    return `[[${target}${pipe === -1 ? "" : inner.slice(pipe)}]]`;
  });
  return raw.slice(0, start) + body;
}

/** Replace the body, keeping the frontmatter block byte-identical. */
export function replaceBody(raw: string, body: string): string {
  const at = frontmatterBlock(raw);
  // No usable block ⇒ the whole file is the body, as `parseNote` reads it.
  return at === null ? body : raw.slice(0, at.bodyStart) + body;
}

File: src/wilcus_vault/note.py
"""Parse and render one note: `---` YAML frontmatter plus a markdown body.

Parsing never raises. A note with broken frontmatter still indexes with the whole
file as its body and `malformed_frontmatter` set for doctor to report.
"""

import hashlib
import json
import re
from dataclasses import dataclass, field
from pathlib import PurePosixPath
from typing import Any

import yaml

# One scanner for both finding links and rewriting them. Single-line and
# length-capped so a stray `[[` cannot pair with a `]]` pages later.
WIKILINK = re.compile(r"\[\[([^\[\]\n]{1,256})\]\]")
_HEADING = re.compile(r"^#[ \t]+(.+?)[ \t]*$", re.MULTILINE)
_CLOSE_FENCE = re.compile(r"^---[ \t]*$", re.MULTILINE)
_NODE_BUDGET = 4096


class _Loader(yaml.SafeLoader):
    """SafeLoader with the YAML 1.2 core schema for plain scalars.

    PyYAML speaks YAML 1.1, where `yes`, `on`, `12:30:00` and `2026-01-02` are
    not strings and `01234` is octal. Obsidian and the previous parser read 1.2,
    so only true/false, decimal/0o/0x integers, floats and null resolve.
    """


_CORE_SCHEMA = {
    "tag:yaml.org,2002:bool": (r"^(?:true|True|TRUE|false|False|FALSE)$", "tTfF"),
    "tag:yaml.org,2002:int": (r"^(?:[-+]?[0-9]+|0o[0-7]+|0x[0-9a-fA-F]+)$", "-+0123456789"),
    "tag:yaml.org,2002:float": (
        r"^(?:[-+]?(?:\.[0-9]+|[0-9]+(?:\.[0-9]*)?)(?:[eE][-+]?[0-9]+)?"
        r"|[-+]?\.(?:inf|Inf|INF)|\.(?:nan|NaN|NAN))$",
        "-+.0123456789",
    ),
}
_Loader.yaml_implicit_resolvers = {
    key: [(tag, rx) for tag, rx in resolvers if tag.endswith((":null", ":merge"))]
    for key, resolvers in _Loader.yaml_implicit_resolvers.items()
}
for _tag, (_pattern, _first) in _CORE_SCHEMA.items():
    _Loader.add_implicit_resolver(_tag, re.compile(_pattern), list(_first))


def _construct_int(loader: yaml.SafeLoader, node: yaml.Node) -> int:
    text = loader.construct_scalar(node)  # type: ignore[arg-type]
    base = 16 if text[:2] == "0x" else 8 if text[:2] == "0o" else 10
    return int(text, base)


_Loader.add_constructor("tag:yaml.org,2002:int", _construct_int)


@dataclass(frozen=True)
class Note:
    path: str  # vault-relative, the note's identity
    slug: str  # filename stem, what a bare `[[stem]]` resolves against
    title: str
    type: str | None
    frontmatter: dict[str, Any]
    body: str
    links: list[str]  # deduped link targets as written, first-seen order
    hash: str  # sha256 hex of the raw text; a dirtiness check, not an identity
    malformed_frontmatter: bool = field(default=False)


def link_target(rel: str) -> str:
    """The path-qualified wikilink target of a vault-relative path: minus `.md`."""
    return rel[: -len(".md")]


def link_of(inner: str) -> str:
    """A match's link target: the text before any `|alias`, trimmed."""
    return inner.split("|")[0].strip()


def is_writable_target(target: str) -> bool:
    """Can this target survive a round trip through `[[...]]`?

    Brackets, pipes, newlines and surrounding whitespace all change what the
    scanner reads back, so a rewrite to such a target is never made.
    """
    return not re.search(r"[\[\]\r\n]", target) and link_of(target) == target


def within_node_budget(value: object, budget: int = _NODE_BUDGET) -> bool:
    """False when the parsed YAML has too many nodes to belong in an index row.

    YAML aliases re-expand on every reference, so a tiny billion-laughs block
    parses into a huge object. Every visit is counted, shared or not.
    """
    seen = 0
    stack = [value]
    while stack:
        seen += 1
        if seen > budget:
            return False
        current = stack.pop()
        if isinstance(current, dict):
            stack.extend(current.values())
        elif isinstance(current, list):
            stack.extend(current)
    return True


def usable_frontmatter(text: str) -> dict[str, Any] | None:
    """The block's YAML as a mapping, or None when it cannot be used as one.

    Every reader of a frontmatter block decides through this one function so
    the parser and the textual patcher can never disagree about what is real.
    """
    if text.strip() == "":
        return {}
    try:
        parsed = yaml.load(text, Loader=_Loader)
    # PyYAML composes and constructs recursively, so nesting deep enough to
    # exhaust the stack raises RecursionError rather than a YAMLError.
    except (yaml.YAMLError, RecursionError):
        return None
    if not isinstance(parsed, dict) or not within_node_budget(parsed):
        return None
    # An explicit `!!binary`, `!!set` or `!!timestamp` builds a value the index
    # cannot store as JSON: the block is malformed, never a crash in reindex.
    try:
        json.dumps(parsed)
    except (TypeError, ValueError):
        return None
    return parsed


def _split_frontmatter(text: str) -> tuple[str | None, str, bool]:
    """(yaml, body, unterminated): split a leading `---` block off the file."""
    if not text.startswith("---\n"):
        return None, text, False
    rest = text[4:]
    close = _CLOSE_FENCE.search(rest)
    if close is None:
        return None, text, True
    body = re.sub(r"^---[ \t]*\n?", "", rest[close.start() :], count=1)
    return rest[: close.start()], body, False


def parse_note(raw: str, rel_path: str) -> Note:
    digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
    text = raw.replace("\r\n", "\n").removeprefix("\ufeff")
    slug = PurePosixPath(rel_path).stem

    yaml_text, split_body, unterminated = _split_frontmatter(text)
    frontmatter: dict[str, Any] = {}
    body = text
    malformed = unterminated
    if yaml_text is not None:
        parsed = usable_frontmatter(yaml_text)
        if parsed is None:
            malformed = True
        else:
            frontmatter, body = parsed, split_body

    fm_title = frontmatter.get("title")
    fm_type = frontmatter.get("type")
    for value in (fm_title, fm_type):
        if value is not None and not isinstance(value, str):
            malformed = True

    # A plain regex scan: links and headings inside code fences count too.
    links = list(dict.fromkeys(link_of(m) for m in WIKILINK.findall(body)))
    heading = _HEADING.search(body)
    title = slug
    if isinstance(fm_title, str) and fm_title.strip():
        title = fm_title
    elif heading is not None:
        title = heading.group(1)

    return Note(
        path=rel_path,
        slug=slug,
        title=title,
        type=fm_type if isinstance(fm_type, str) else None,
        frontmatter=frontmatter,
        body=body,
        links=[link for link in links if link],
        hash=digest,
        malformed_frontmatter=malformed,
    )


def serialize_note(frontmatter: dict[str, Any], body: str) -> str:
    """Render a note we authored. Never use it to rewrite a human's file: a YAML
    round-trip drops comments and rewrites `01234` and `1.0`."""
    if not frontmatter:
        return body
    # One line per key, however long the value: the textual patcher edits by line.
    dumped = yaml.safe_dump(frontmatter, sort_keys=False, allow_unicode=True, width=2**31)
    return f"---\n{dumped.rstrip('\n')}\n---\n{body}"
